src/scheduler/MpiThreadPool.cpp

```diff
@@ -0,0 +1,98 @@
+#include <faabric/scheduler/MpiThreadPool.h>
+#include <faabric/util/logging.h>
+
+namespace faabric::scheduler {
+MpiAsyncThreadPool::MpiAsyncThreadPool(int nThreads)
+  : shutdown(false)
+{
+    faabric::util::getLogger()->debug(
+      "Starting an MpiAsyncThreadPool of size {}", nThreads);
+
+    // Initialize async. req queue
+    localReqQueue = std::make_shared<MpiReqQueue>();
+
+    // Initialize thread pool
+    for (int i = 0; i < nThreads; ++i) {
+        threadPool.emplace_back(
+          std::bind(&MpiAsyncThreadPool::entrypoint, this, i));
+    }
+}
+
+MpiAsyncThreadPool::~MpiAsyncThreadPool()
+{
+
+    faabric::util::getLogger()->debug("Shutting down MpiAsyncThreadPool");
+    this->shutdown = true;
+
+    // When the destructor is called, either some thread will have hit the
+    // dequeue timeout and won't be joinable, or none has and all are joinable.
+    // We wait untill all hit the timeout, saving an additional cond. variable.
+    for (auto& thread : threadPool) {
+        if (thread.joinable()) {
+            thread.join();
+        }
+    }
+}
+
+void MpiAsyncThreadPool::awaitAsyncRequest(int reqId)
+{
+    faabric::util::UniqueLock lock(awakeMutex);
+
+    // Wait until the condition is signaled and the predicate is false.
+    // Note that the same lock protects concurrent accesses to finishedReqs
+    awakeCV.wait(lock, [this, reqId] {
+        return this->finishedReqs.find(reqId) != this->finishedReqs.end();
+    });
+
+    // Before giving up the lock, remove our request as it has already finished
+    auto it = finishedReqs.find(reqId);
+    if (it == finishedReqs.end()) {
+        throw std::runtime_error(
+          fmt::format("Error: unrecognized reqId {}", reqId));
+    }
+    finishedReqs.erase(it);
```

Comment:
How come we can't just call `finishedReqs.erase(reqId)`?

Reply:
I double-check the iterator (for sanity) and given that I already have it, it's faster to erase from the iterator (as we don't have to look it up again).